pkg/api/vm/store.go

```diff
@@ -0,0 +1,101 @@
+package vm
+
+import (
+	"fmt"
+
+	"github.com/rancher/apiserver/pkg/apierror"
+	"github.com/rancher/apiserver/pkg/types"
+	"github.com/rancher/wrangler/pkg/schemas/validation"
+	"github.com/rancher/wrangler/pkg/slice"
+	k8sapierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	ctlcdiv1beta1 "github.com/rancher/harvester/pkg/generated/controllers/cdi.kubevirt.io/v1beta1"
+	ctlkubevirtv1alpha3 "github.com/rancher/harvester/pkg/generated/controllers/kubevirt.io/v1alpha3"
+)
+
+type vmStore struct {
+	types.Store
+
+	vmCache     ctlkubevirtv1alpha3.VirtualMachineCache
+	dataVolumes ctlcdiv1beta1.DataVolumeClient
+}
+
+func (s *vmStore) Delete(request *types.APIRequest, schema *types.APISchema, id string) (types.APIObject, error) {
+	deleteDisks := request.Query["deleteDisks"]
+	vm, err := s.vmCache.Get(request.Namespace, request.Name)
+	if err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, fmt.Sprintf("Failed to get vm %s/%s, %v", request.Namespace, request.Name, err))
+	}
+
+	var keepDisks []string
+	var keepDataVolumes []string
+	var deleteDataVolume []string
+	if vm.Spec.Template != nil {
+		for _, v := range vm.Spec.Template.Spec.Domain.Devices.Disks {
+			if !slice.ContainsString(deleteDisks, v.Name) {
+				keepDisks = append(keepDisks, v.Name)
+			}
+		}
+
+		for _, vol := range vm.Spec.Template.Spec.Volumes {
+			if slice.ContainsString(keepDisks, vol.Name) && vol.DataVolume != nil {
+				keepDataVolumes = append(keepDataVolumes, vol.DataVolume.Name)
+			}
+
+			if slice.ContainsString(deleteDisks, vol.Name) && vol.DataVolume != nil {
+				deleteDataVolume = append(deleteDataVolume, vol.DataVolume.Name)
+			}
+		}
+	}
+
+	if err := s.updateDataVolumeOwnerRef(vm.Namespace, vm.Name, keepDataVolumes); err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, fmt.Sprintf("Failed to update OwnerReferences for vm %s/%s, %v", request.Namespace, request.Name, err))
+	}
+
+	apiObj, err := s.Store.Delete(request, request.Schema, id)
+	if err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, fmt.Sprintf("Failed to delete vm %s/%s, %v", request.Namespace, request.Name, err))
+	}
+
+	if err := s.deleteDataVolumes(vm.Namespace, deleteDataVolume); err != nil {
+		return types.APIObject{}, apierror.NewAPIError(validation.ServerError, fmt.Sprintf("Failed to delete dataVolume for vm %s/%s, %v", request.Namespace, request.Name, err))
+	}
+	return apiObj, nil
+}
+
+func (s *vmStore) updateDataVolumeOwnerRef(vmNamespace, vmName string, keepDataVolumes []string) error {
```

Comment:
prefer to call it `removeVMDataVolumeOwnerRef` since this is what the method does.